temporalio/converter.py

```diff
@@ -0,0 +1,410 @@
+
+    @property
+    def encoding(self) -> str:
+        """See base class."""
+        return "binary/null"
+
+    async def encode(self, value: Any) -> Optional[temporalio.api.common.v1.Payload]:
+        """See base class."""
+        if value is None:
+            return temporalio.api.common.v1.Payload(
+                metadata={"encoding": self.encoding.encode()}
+            )
+        return None
+
+    async def decode(
+        self,
+        payload: temporalio.api.common.v1.Payload,
+        type_hint: Optional[Type] = None,
+    ) -> Any:
+        if len(payload.data) > 0:
+            raise RuntimeError("Expected empty data set for binary/null")
+        return None
+
+
+class BinaryPlainPayloadConverter(PayloadConverter):
+    """Converter for 'binary/plain' payloads supporting bytes values."""
+
+    @property
+    def encoding(self) -> str:
+        """See base class."""
+        return "binary/plain"
+
+    async def encode(self, value: Any) -> Optional[temporalio.api.common.v1.Payload]:
+        """See base class."""
+        if isinstance(value, bytes):
+            return temporalio.api.common.v1.Payload(
+                metadata={"encoding": self.encoding.encode()}, data=value
+            )
+        return None
+
+    async def decode(
+        self,
+        payload: temporalio.api.common.v1.Payload,
+        type_hint: Optional[Type] = None,
+    ) -> Any:
+        """See base class."""
+        return payload.data
+
+
+_sym_db = google.protobuf.symbol_database.Default()
+
+
+class JSONProtoPayloadConverter(PayloadConverter):
+    """Converter for 'json/protobuf' payloads supporting protobuf Message values."""
+
+    @property
+    def encoding(self) -> str:
+        """See base class."""
+        return "json/protobuf"
+
+    async def encode(self, value: Any) -> Optional[temporalio.api.common.v1.Payload]:
+        """See base class."""
+        if isinstance(value, google.protobuf.message.Message):
+            # We have to convert to dict then to JSON because MessageToJson does
+            # not have a compact option removing spaces and newlines
+            json_str = json.dumps(
+                google.protobuf.json_format.MessageToDict(value),
+                separators=(",", ":"),
+                sort_keys=True,
+            )
+            return temporalio.api.common.v1.Payload(
+                metadata={
+                    "encoding": self.encoding.encode(),
+                    "messageType": value.DESCRIPTOR.full_name.encode(),
+                },
+                data=json_str.encode(),
+            )
+        return None
+
+    async def decode(
+        self,
+        payload: temporalio.api.common.v1.Payload,
+        type_hint: Optional[Type] = None,
+    ) -> Any:
+        """See base class."""
+        message_type = payload.metadata.get("messageType", b"<unknown>").decode()
+        try:
+            value = _sym_db.GetSymbol(message_type)()
```

Comment:
Since we have the type hint here I recommend using that and falling back to the messageType header.

Reply:
The type hint is optional for proto payloads, the message type is not (at least how we're developing this across SDKs), so we might as well use the thing we know is always there.

Reply:
I know Dmitry had his concerns with breaking compatibility if the message name changes.

Reply:
If we're concerned with message name changes in the payload, he should also be concerned with incompatible type hint changes. As it stands, there are changes a dev can make that can break things.